### ISSUE TITLE
Magento 2.4.2 CLS issue

### ISSUE THREAD

**Issue by gondas33** [magento/magento2#32956]
As you might know, Google recently introduced a new ranking factor and that is CLS - Cumulative Layout Shift.
I have done few tests on default magento 2 stores and each of them fails the test of having 0 CLS on a product page.

Having a higher CLS than 0.25 as page will slowly destroy your organic rankings as also fail the test called 'Core Vitals' in Google Search Console.

### Preconditions (*)

1. Magento 2.4.2
2. With Varnish/Without varnish

### Steps to reproduce (*)

1. https://developers.google.com/speed/pagespeed/insights/ on any product page
2. Gets a CLS score of 0.47 or higher
3. Fails Core Vitals test in Search Console

### Expected result (*)

1. The test results of the CLS should be 0 or less than 0.25

### Actual result (*)

1. CLS 0.47 or higher

**Below is more information to reproduce issue step by step:**
1. Create a simple product having a attribute set options (including navigation bar layout enabled)
2. Also make sure to add a few images and one vimeo video url with base option provided
3. Saved product should be visible in the frond end.
4. Navigate to front end and open the same product page.
5. Check the CLS score from the browser's developer tool.
6. Below is the screenshot ie resulted with 0.6844
![image](https://user-images.githubusercontent.com/51679138/130939034-eaf150c1-ee68-4bb9-95ed-a69b6ce313be.png)

7. Expected to be <=0.25

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.4/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ X ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by an4rei** [comment]
@magento give me 2.4-develop instance

**Comment by an4rei** [comment]
Issue exists on latest version as well:
![image](https://user-images.githubusercontent.com/11984860/117210847-e3cc9280-ae00-11eb-93c6-f9301b17110f.png)

**Comment by ihor-sviziev** [comment]
@sivaschenko @gabrieldagama, I think priority should be set to P2 for this issue
@mrtuvn, you might be interested in fixing this issue :)

**Comment by mrtuvn** [comment]
Is this issue remain in latest branch or official 2.4.2 ? In my opinion this issue should have P3-P4 priority 
More precisely title should be "Improvement CLS in product details magento" Since we have a lot of page maybe next one will fix for category

**Comment by mrtuvn** [comment]
![Screenshot from 2021-05-06 19-25-29](https://user-images.githubusercontent.com/1908873/117297872-ea124b80-aea0-11eb-9d1c-388fd064c496.png)
Here is my test in latest code base
Detail page luma setup sampledata (incognito mode)

**Comment by ihor-sviziev** [comment]
@mrtuvn, unfortunately, pagespeed usually shows different results than on localhost. So pls check it on some server

**Comment by mrtuvn** [comment]
Sorry i dont have any live server currently for that. But i think results may not difference too much. Yes we still need more improve in mobile when cls metric in mobile not good for user

**Comment by ihor-sviziev** [comment]
@mrtuvn you can check it on test instance

**Comment by onlinebizsoft** [comment]
I would like to see the result and see how this being improved

**Comment by mrtuvn** [comment]
Since details page magento have fotorama media image may cause adjustment when first time load (since image not have attr width height). Also i notices swatch opt load later also may cause cls problem too

**Comment by chornobylskyi** [comment]
@magento I am working on this

**Comment by chornobylskyi** [comment]
@magento add to contributors team

**Comment by chornobylskyi** [comment]
@magento I am working on this

**Comment by engcom-Alfa** [comment]
I have rested in 4.2-develop version and updated the description with more information. Since it is reproducible, confirming the issue.

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-Alfa 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-42599` were created

**Issue Available**: @engcom-Alfa, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by tpaaro** [comment]
For anyone struggling with PDP CLS issue, this is what I did to solve it on my `2.3.5-p2` Magento CE instance.
1. Defined width and height for the placeholder image, just as it is done in this [PR](https://github.com/magento/magento2/pull/33265).
    * For responsive resizing I also added this:
```less
// app/design/frontend/<Vendor>/<theme>/Magento_Catalog/web/css/source/_extend.less
& when (@media-common = true) {
    .gallery-placeholder__image {
        height: auto;
    }
}
```
2. Updated [gallery.js](https://github.com/magento/magento2/blob/2.4-develop/lib/web/mage/gallery/gallery.js) to the latest version.
    * `app/design/frontend/<Vendor>/<theme>/web/mage/gallery/gallery.js`
    * For me, this also fixed the [on load blinking issue](https://github.com/magento/magento2/issues/28339), too
3. Moved `product.info.main` block after `product.info.media` block:
    * Reason for this is to render the gallery's placeholder image before the main content. Otherwise there will be CLS happening that will take place for a fraction of a second (very hard to detect visually, but not impossible)
```xml
// app/design/frontend/<Vendor>/<theme>/Magento_Catalog/layout/catalog_product_view.xml
<?xml version="1.0"?>
<page xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:View/Layout/etc/page_configuration.xsd">
    <body>
        <move element="product.info.main" destination="content" after="product.info.media" />
    </body>
</page>
```
4. Last, but not least, the Fotorama's navigation dots - once the gallery loads they shift the main content by 30px. My solution was to use `position: absolute` so they would not do any shifting.
    * Caution: this makes the dots to appear on the image itself so for some this may not be an acceptable solution.
 ```less
// app/design/frontend/<Vendor>/<theme>/web/css/source/_extend.less

// CLS fix
.fotorama__nav-wrap {
    position: absolute;
    bottom: 0;
}

// optional - make dots bigger
.fotorama__dot {
    height: 8px;
    width: 8px;
}
```  

After these changes I managed to bring CLS down to **0.026** from **0.35**.

**Comment by tpaaro** [comment]
I wasn't pleased with the `position: absolute` solution so I had another go.

The new approach is about moving Fotorama's navigation wrapper to another element that has a **fixed height**. Moving takes place once the gallery has been loaded.

**Full changelog:**
```diff
diff --git a/app/design/frontend/<Vendor>/<theme>/Magento_Catalog/templates/product/view/gallery.phtml b/app/design/frontend/<Vendor>/<theme>/Magento_Catalog/templates/product/view/gallery.phtml
index 9d0212f2f..29ca2092f 100644
--- a/app/design/frontend/<Vendor>/<theme>/Magento_Catalog/templates/product/view/gallery.phtml
+++ b/app/design/frontend/<Vendor>/<theme>/Magento_Catalog/templates/product/view/gallery.phtml
@@ -50,6 +50,7 @@ if ($mainImage) {
         <?= $imageHeight ? 'height="'. $block->escapeHtmlAttr($imageHeight) .'"' : '' ?>
     />
 </div>
+<div class="gallery__nav-container"></div>
 
 <script type="text/x-magento-init">
     {
diff --git a/app/design/frontend/<Vendor>/<theme>/web/css/source/_extend.less b/app/design/frontend/<Vendor>/<theme>/web/css/source/_extend.less
index 31befdf27..b3fb10b4a 100644
--- a/app/design/frontend/<Vendor>/<theme>/web/css/source/_extend.less
+++ b/app/design/frontend/<Vendor>/<theme>/web/css/source/_extend.less
@@ -55,7 +55,7 @@
 @button-primary__active__border: @button-primary__border;
 @button-primary__active__color: @button-primary__color;
 
-& when (@media-common = true) {
+& when (@media-common =true) {
     .d-none {
         display: none !important;
     }
@@ -76,12 +76,13 @@
     }
 }
 
-.media-width(@extremum, @break) when (@extremum ='max') and (@break = @screen__m) {
-    // CLS fix
-    .fotorama__nav-wrap {
-        bottom: 0;
-        position: absolute;
+.media-width(@extremum, @break) when (@extremum ='max') and (@break =@screen__m) {
+
+    // START: CLS fix
+    .gallery__nav-container {
+        height: 30px;
     }
+    // END: CLS fix
 
     .fotorama__active {
         .fotorama__dot {
@@ -93,4 +94,4 @@
             top: 10px;
         }
     }
-}
+}
\ No newline at end of file
diff --git a/app/design/frontend/<Vendor>/<theme>/web/mage/gallery/gallery.js b/app/design/frontend/<Vendor>/<theme>/web/mage/gallery/gallery.js
index da47bb36a..1318ddb6a 100644
--- a/app/design/frontend/<Vendor>/<theme>/web/mage/gallery/gallery.js
+++ b/app/design/frontend/<Vendor>/<theme>/web/mage/gallery/gallery.js
@@ -270,6 +270,7 @@
                     next: $t('Next'),
                     previous: $t('Previous')
                 }),
+                isTouchEnabled = this.isTouchEnabled,
                 mainImageIndex,
                 $element = settings.$element,
                 $fotoramaElement,
@@ -316,6 +317,14 @@
                             .css('min-height', '');
 
                         $fotoramaStage.css('position', '');
+                        // START: CLS fix
+                        if (isTouchEnabled) {
+                            const navWrap = $element.find('.fotorama__nav-wrap').detach();
+                            navWrap.removeClass('d-none');
+
+                            $element.next('.gallery__nav-container').append(navWrap);
+                        }
+                        // END: CLS fix
                     });
             settings.$elementF = $fotoramaElement;
             settings.fotoramaApi = $fotoramaElement.data('fotorama');
diff --git a/lib/web/mage/gallery/gallery.html b/lib/web/mage/gallery/gallery.html
index a258ede20..51b05e5c1 100644
--- a/lib/web/mage/gallery/gallery.html
+++ b/lib/web/mage/gallery/gallery.html
@@ -22,7 +22,7 @@
             <div class="fotorama__zoom-out" data-gallery-role="fotorama__zoom-out" aria-label="Zoom out" role="button" tabindex="0"></div>
             <div class="fotorama__spinner"></div>
         </div>
-        <div class="fotorama__nav-wrap" data-gallery-role="nav-wrap">
+        <div class="fotorama__nav-wrap d-none" data-gallery-role="nav-wrap">
             <div class="fotorama__nav fotorama__nav--thumbs">
                 <div class="fotorama__fullscreen-icon" data-gallery-role="fotorama__fullscreen-icon" tabindex="0" aria-label="Exit fullscreen" role="button"></div>
                 <div class="fotorama__thumb__arr fotorama__thumb__arr--left" role="button" aria-label="Previous" data-gallery-role="arrow" tabindex = "-1">
```

Note: `d-none` adds `display: none !important`.

**Comment by tpaaro** [comment]
I was not satisfied with the `position: absolute` solution so I had another go.

New approach is about moving Fotorama's navigation wrapper element to a **fixed height** container.

**Full changelog:**
```diff
diff --git a/app/design/frontend/<Vendor>/<theme>/Magento_Catalog/templates/product/view/gallery.phtml b/app/design/frontend/<Vendor>/<theme>/Magento_Catalog/templates/product/view/gallery.phtml
index 9d0212f2f..29ca2092f 100644
--- a/app/design/frontend/<Vendor>/<theme>/Magento_Catalog/templates/product/view/gallery.phtml
+++ b/app/design/frontend/<Vendor>/<theme>/Magento_Catalog/templates/product/view/gallery.phtml
@@ -50,6 +50,7 @@ if ($mainImage) {
         <?= $imageHeight ? 'height="'. $block->escapeHtmlAttr($imageHeight) .'"' : '' ?>
     />
 </div>
+<div class="gallery__nav-container"></div>
 
 <script type="text/x-magento-init">
     {
diff --git a/app/design/frontend/<Vendor>/<theme>/web/css/source/_extend.less b/app/design/frontend/<Vendor>/<theme>/web/css/source/_extend.less
index 31befdf27..b3fb10b4a 100644
--- a/app/design/frontend/<Vendor>/<theme>/web/css/source/_extend.less
+++ b/app/design/frontend/<Vendor>/<theme>/web/css/source/_extend.less
@@ -76,12 +76,13 @@
     }
 }
 
-.media-width(@extremum, @break) when (@extremum ='max') and (@break = @screen__m) {
-    // CLS fix
-    .fotorama__nav-wrap {
-        bottom: 0;
-        position: absolute;
+.media-width(@extremum, @break) when (@extremum ='max') and (@break =@screen__m) {
+
+    // START: CLS fix
+    .gallery__nav-container {
+        height: 30px;
     }
+    // END: CLS fix
 
     .fotorama__active {
         .fotorama__dot {
@@ -93,4 +94,4 @@
             top: 10px;
         }
     }
-}
+}
\ No newline at end of file
diff --git a/app/design/frontend/<Vendor>/<theme>/web/mage/gallery/gallery.js b/app/design/frontend/<Vendor>/<theme>/web/mage/gallery/gallery.js
index da47bb36a..55e2d39c9 100644
--- a/app/design/frontend/<Vendor>/<theme>/web/mage/gallery/gallery.js
+++ b/app/design/frontend/<Vendor>/<theme>/web/mage/gallery/gallery.js
@@ -316,6 +316,22 @@
                             .css('min-height', '');
 
                         $fotoramaStage.css('position', '');
+                        // START: CLS fix
+                        let navWrap = $element.find(".fotorama__nav-wrap"),
+                            isMobile = matchMedia(
+                                "only screen and (max-width: 767px)"
+                            ).matches;
+                        if (isMobile) {
+                            navWrap = navWrap.detach();
+                            navWrap.removeClass("d-none");
+
+                            $element
+                                .next(".gallery__nav-container")
+                                .append(navWrap);
+                        } else {
+                            navWrap.removeClass("d-none");
+                        }
+                        // END: CLS fix
                     });
             settings.$elementF = $fotoramaElement;
             settings.fotoramaApi = $fotoramaElement.data('fotorama');
diff --git a/lib/web/mage/gallery/gallery.html b/lib/web/mage/gallery/gallery.html
index a258ede20..51b05e5c1 100644
--- a/lib/web/mage/gallery/gallery.html
+++ b/lib/web/mage/gallery/gallery.html
@@ -22,7 +22,7 @@
             <div class="fotorama__zoom-out" data-gallery-role="fotorama__zoom-out" aria-label="Zoom out" role="button" tabindex="0"></div>
             <div class="fotorama__spinner"></div>
         </div>
-        <div class="fotorama__nav-wrap" data-gallery-role="nav-wrap">
+        <div class="fotorama__nav-wrap d-none" data-gallery-role="nav-wrap">
             <div class="fotorama__nav fotorama__nav--thumbs">
                 <div class="fotorama__fullscreen-icon" data-gallery-role="fotorama__fullscreen-icon" tabindex="0" aria-label="Exit fullscreen" role="button"></div>
                 <div class="fotorama__thumb__arr fotorama__thumb__arr--left" role="button" aria-label="Previous" data-gallery-role="arrow" tabindex = "-1">
```

**Comment by nkajic** [comment]
Hi @ gondas33 thanks for this fix. It works perfectly on 2.4.3-p3. Is solved my CLS from 0.36 on mobile to 0.00! Same it is not merged yet.

**Comment by aceopticsuk** [comment]
Hi, I am currently running Magento 2.4.5-p1 and having lots of CLS issues after installing a new theme. I tried to apply some of these settings but I couldn't find many of the files that have to be edited, so I was wondering if some file routes changed in Magento 2.4.5. For example, under app/code there is no Magento folder, so there is no way to find the file catalog_product_view.xml (this repeats with several files). Any help or orientation would be appreciated.

**Comment by IbrahimS2** [comment]
@aceopticsuk Share your instance URL, and are you using the luma theme?

**Comment by aceopticsuk** [comment]
> @aceopticsuk Share your instance URL, and are you using the luma theme?

Hi there, I was speaking about our live site https://aceoptics.co.uk/ - I am new to GitHub and have been using it just to read articles and solutions to errors, that was actually my first comment ever. Some weeks ago we installed the Pearl Theme on our ecommerce store, before we used a customized child theme having Luma as parent theme, and we didn't have severe CLS issues until now.

**Comment by IbrahimS2** [comment]
@aceopticsuk I worked with the WP Pearl theme, it is good but fully optimized, you have CLS issues that must be addressed, I suggest the following:
- Assigning fixed height/width to the images to reduce the CLS, I also suggest using picture tags to serve appropriately sized images
- And product gallery you will need to override by copying core files into your child template and applying the fix from the previous comments
- If you wish to have this professionally looked at, then you better send me a pm

**Comment by aceopticsuk** [comment]
> @aceopticsuk I worked with the WP Pearl theme, it is good but not fully optimized, you have CLS issues that must be addressed, I suggest the following:
> 
> * Assigning fixed height/width to the images to reduce the CLS, I also suggest using picture tags to serve appropriately sized images
> * And product gallery you will need to override by copying core files into your child template and applying the fix from the previous comments
> * If you wish to have this professionally looked at, then you better send me a pm

Thanks. I will see what I can do. But I thought this workaround would work for our Magento site

**Comment by ben-hgc** [comment]
Hi everyone, thank you @tpaaro I have managed to get down the CLS score slightly on Magento 2.4.4 by following your advice.
I carried out the following points:
2. Updated [gallery.js](https://github.com/magento/magento2/blob/2.4-develop/lib/web/mage/gallery/gallery.js) to the latest version.

3. Moved product.info.main block after product.info.media block:

4. Allowed enough space 412px for Fotorama's navigation dots.

However I'm still getting a bad CLS score on Mobile when there are Fotorama gallery images such as on this page:
[https://www.homeandgardencentre.co.uk/garden-furniture/kettler/elba/kettler-elba-low-corner-lounge-set-with-teak-top-coffee-table.html](url)
![CLS_Score](https://github.com/magento/magento2/assets/138235397/3a36dd7e-729f-4a96-bd9a-d8e27b136523)

I can't figure out how this is happening, as on the page load nothing appears to shift at all.  It is only when a product has multiple images and therefore a Fotorama gallery, the gallery dots also don't shift due to the 412px space allowed for the gallery dots.

Specifically what did you carry out from your point 1. in your post to resolve the CLS issue on your website?  Or what else do you think I'm missing?
1. Defined width and height for the placeholder image, just as it is done in this https://github.com/magento/magento2/pull/33265.

Thank you in advance,
Ben

**Comment by ihor-sviziev** [comment]
The issue is fixed https://github.com/magento/magento2/pull/33265 (merged in https://github.com/magento/magento2/commit/b9f57bda3c44f4320f966a67c128075ae830ca42)

**Comment by sdouma** [comment]
@magento give me 2.4-develop instance
